### PR TITLE
Add the -o flag to 1.6 dnsmasq startup.

### LINKED ISSUE #2546: DNS behaving oddly when using split-horizon nameservers in /etc/resolv.conf
We have a 1.6 cluster set up using [the newest `kube-dns` configuration](https://github.com/kubernetes/kops/blob/6201e924c7559148cdd10e54a11c6479a9e95a60/upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template).

When we run this in AWS, we see occasional DNS failures when looking up internal domains. I have traced this to our `/etc/resolv.conf` file, which looks something like:

```
nameserver 10.16.0.4
nameserver 8.8.8.8
search internal-name us-west-2.compute.internal
```
This is configured via DCHP in our subnet  (we're using a shared subnet for our cluster).

The problem we're having is that `dnsmasq` by default uses a nondeterministic `/etc/resolv.conf` server to perform lookups - which is NOT the behavior used in `libc`. Because of this, we'll get occasional failures when looking up internal-only URLs - and worse yet, `dnsmasq` will *also* cache the negative responses from `8.8.8.8`, meaning we'll eventually get into a state where our DNS mostly won't resolve.

`dnsmasq` has a flag (`-o`) that makes it behave more like `libc`, in that it will try the nameservers in-order, instead of randomly. I have a patch for this I'll submit in a moment.

(I first reported this issue as https://github.com/kubernetes/dns/issues/85, before I realized this was related to the configuration used).

### PR DESCRIPTION
**Pull request by jkinkead**
Fixes #2546 .

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2547)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

File: upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template [hunk 1]
```diff
@@ -162,6 +162,7 @@ spec:
         - -restartDnsmasq=true
         - --
         - -k
+        - -o
```
Comment (chrislovecnm) [inline upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template]: Does this exist in pre 1.6?

Comment (jkinkead) [inline upup/models/cloudup/resources/addons/kube-dns.addons.k8s.io/k8s-1.6.yaml.template]: The flag does not. `dnsmasq` is run directly (not through `dnsmasq-nanny`), and has a fairly different configuration. I can't say for certain sure it's needed, but it probably is.

Would you like me to add the `-o` flag there as well?

**Comment by jkinkead** [comment]
So, this worked for a few hours - quite a bit better than the previous setup - but still ended up getting flaky after a while. I'm not sure why.

I'm going to rework our setup to use [the new `stubDomains` feature](http://blog.kubernetes.io/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes.html). This should theoretically work 100% correctly.

This could still be merged. Let me know what you think!

**Comment by justinsb** [comment]
I commented on the upstream issue https://github.com/kubernetes/dns/issues/85

It seems to me like `-o` is the correct solution?  But, because this isn't in the upstream configuration, I'm wary of introducing it unilaterally.

BTW @jkinkead I can't reopen https://github.com/kubernetes/dns/issues/85, but maybe you can?

**Comment by justinsb** [comment]
Based on bowei's comment here https://github.com/kubernetes/dns/issues/85#issuecomment-301224997, it sounds like the answer is not to have split horizon DNS servers.  So you shouldn't have 8.8.8.8 as a DNS server if you also use the AWS VPC DNS server.

Is this an option for you @jkinkead ?  Is the a requirement to have both 8.8.8.8 and the AWS DNS?

**Comment by justinsb** [comment]
Going to remove `blocks-next` for now

**Comment by jkinkead** [comment]
Thanks for the feedback. I don't think the DNS configuration we have is wrong per se, since [libc is documented as querying servers in-order](http://man7.org/linux/man-pages/man5/resolv.conf.5.html), but I also think that relying on `kube-dns` interacting with `/etc/resolv.conf` for private DNS is not recommended in general.

Since `-o` doesn't seem to be reliable - or the recommended solution from the `kube-dns` team - I'll close this PR out.

I think the issue should likely stay open, since it's surprising behavior. Maybe the default should be to totally ignore `/etc/resolv.conf`, and require DNS configuration to fully live in the ConfigMap for `kube-dns`?

**Pull request closed without merging**